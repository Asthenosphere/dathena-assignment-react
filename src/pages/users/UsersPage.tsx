import {
  IonAlert,
  IonButtons,
  IonContent,
  IonFab,
  IonFabButton,
  IonHeader,
  IonIcon,
  IonItem,
  IonLabel,
  IonList,
  IonLoading,
  IonMenuButton,
  IonPage,
  IonText,
  IonTitle,
  IonToolbar,
} from "@ionic/react";
import React, { useEffect, useReducer } from "react";
import moment from "moment";
import { User } from "../../interfaces/User";
import { getUsers } from "../../services/userService";
import "./UsersPage.css";
import { addOutline } from "ionicons/icons";
import Container from "../../components/Container";
import UserModal from "../../components/UserModal";
import UserForm from "../../components/UserForm";

interface UsersPageState {
  users: User[];
  currentUser: User | null;
  isAlertVisible: boolean;
  hasConfirm: boolean;
  alertHeader: string;
  alertMessage: string;
  alertHandler: () => void;
  confirmHandler: () => void;
  isUserModalVisible: boolean;
  isUserFormVisible: boolean;
  isLoading: boolean;
}

const UsersPage: React.FC = () => {
  const [state, setState] = useReducer(
    (s: UsersPageState, a: Partial<UsersPageState>) => ({
      ...s,
      ...a,
    }),
    {
      users: [],
      currentUser: null,
      isAlertVisible: false,
      hasConfirm: false,
      alertHeader: "",
      alertMessage: "",
      alertHandler: () => {
        setState({ isAlertVisible: false });
      },
      confirmHandler: () => {
        setState({ isAlertVisible: false });
      },
      isUserModalVisible: false,
      isUserFormVisible: false,
      isLoading: true,
    }
  );

  const fetchData = async () => {
    try {
      const allUsers = await getUsers();
      await setState({ users: allUsers });
      setState({ isLoading: false });
    } catch (error) {
      setState({
        alertHeader: "Error",
        alertMessage: "There was a problem fetching the users.",
        isAlertVisible: true,
        alertHandler: () => {
          setState({ isAlertVisible: false });
        },
      });
    }
  };

  const userCallback = () => {
    fetchData();
  };

  const loadingCallback = (status: boolean) => {
    setState({ isLoading: status });
  };

  const alertCallback = (
    header: string,
    message: string,
    hasConfirm: boolean,
    confirmHandler: () => void
  ) => {
    setState({
      alertHeader: header,
      alertMessage: message,
      hasConfirm: hasConfirm,
      isAlertVisible: true,
      confirmHandler: confirmHandler,
    });
  };

  useEffect(() => {
    fetchData();
  }, []);

  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonButtons slot='start'>
            <IonMenuButton />
          </IonButtons>
          <IonTitle>Users</IonTitle>
        </IonToolbar>
      </IonHeader>

      <IonContent fullscreen>
        <IonHeader collapse='condense'>
          <IonToolbar>
            <IonTitle size='large'>Users</IonTitle>
          </IonToolbar>
        </IonHeader>
        {state.currentUser && (
          <UserModal
            key={state.currentUser.id}
            user={state.currentUser!}
            isUserModalVisible={state.isUserModalVisible}
            setIsUserModalVisible={(status: boolean) => {
              setState({ isUserModalVisible: status });
            }}
            userCallback={userCallback}
            loadingCallback={loadingCallback}
            alertCallback={alertCallback}
          />
        )}
        <UserForm
          users={state.users}
          isUserFormVisible={state.isUserFormVisible}
          setIsUserFormVisible={(status: boolean) => {
            setState({ isUserFormVisible: status });
          }}
          userCallback={userCallback}
          loadingCallback={loadingCallback}
          alertCallback={alertCallback}
        />
        {state.users.length > 0 ? (
          <IonList>
            {state.users.map((user, index) => {
              return (
                <IonItem
                  key={index}
                  button
                  color={index % 2 === 0 ? "light" : undefined}
                  onClick={() => {
                    setState({ currentUser: user, isUserModalVisible: true });
                  }}
                >
                  <IonLabel>
                    <h2 style={{ fontWeight: "bold" }}>
                      {user.firstName + " " + user.lastName}
                    </h2>
                    <h3>{"Email: " + user.email}</h3>
                    <h3>
                      {"Date of Birth: " +
                        moment(user.dob!).format("DD MMM YYYY")}
                    </h3>
                  </IonLabel>
                </IonItem>
              );
            })}
          </IonList>
        ) : (
          <Container>
            <IonText>No users found</IonText>
          </Container>
        )}
        <IonFab vertical='bottom' horizontal='end' slot='fixed'>
          <IonFabButton
            color='secondary'
            onClick={() => setState({ isUserFormVisible: true })}
          >
            <IonIcon icon={addOutline} />
          </IonFabButton>
        </IonFab>
        <IonLoading
          isOpen={state.isLoading}
          onDidDismiss={() => {
            setState({ isLoading: false });
          }}
          message={"Loading..."}
          duration={8000}
          spinner='crescent'
        />
        <IonAlert
          isOpen={state.isAlertVisible}
          header={state.alertHeader}
          message={state.alertMessage}
          backdropDismiss={false}
          buttons={
            state.hasConfirm
              ? [
                  {
                    text: "Cancel",
                    role: "cancel",
                    handler: state.alertHandler,
                  },
                  {
                    text: "Proceed",
                    handler: () => {
                      state.confirmHandler();
                      setState({ isAlertVisible: false });
                    },
                  },
                ]
              : [
                  {
                    text: "OK",
                    handler: () => {
                      state.alertHandler();
                      setState({ isAlertVisible: false });
                    },
                  },
                ]
          }
        />
      </IonContent>
    </IonPage>
  );
};

export default UsersPage;
