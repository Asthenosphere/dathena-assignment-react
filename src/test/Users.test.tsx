import React from "react";
import UsersPage from "../pages/users/UsersPage";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";

configure({ adapter: new Adapter() });

describe("users component", () => {
  it("should render users correctly", () => {
    const component = shallow(<UsersPage />);
    expect(component.getElements()).toMatchSnapshot();
  });
});
