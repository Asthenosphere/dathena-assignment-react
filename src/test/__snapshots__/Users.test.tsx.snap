// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users component should render users correctly 1`] = `
Array [
  <ForwardRef(IonPage)>
    <ForwardRef(IonHeader)>
      <ForwardRef(IonToolbar)>
        <ForwardRef(IonButtons)
          slot="start"
        >
          <ForwardRef(IonMenuButton) />
        </ForwardRef(IonButtons)>
        <ForwardRef(IonTitle)>
          Users
        </ForwardRef(IonTitle)>
      </ForwardRef(IonToolbar)>
    </ForwardRef(IonHeader)>
    <ForwardRef(IonContent)
      fullscreen={true}
    >
      <ForwardRef(IonHeader)
        collapse="condense"
      >
        <ForwardRef(IonToolbar)>
          <ForwardRef(IonTitle)
            size="large"
          >
            Users
          </ForwardRef(IonTitle)>
        </ForwardRef(IonToolbar)>
      </ForwardRef(IonHeader)>
      <UserForm
        alertCallback={[Function]}
        isUserFormVisible={false}
        loadingCallback={[Function]}
        setIsUserFormVisible={[Function]}
        userCallback={[Function]}
        users={Array []}
      />
      <Container>
        <ForwardRef(IonText)>
          No users found
        </ForwardRef(IonText)>
      </Container>
      <ForwardRef(IonFab)
        horizontal="end"
        slot="fixed"
        vertical="bottom"
      >
        <ForwardRef(IonFabButton)
          color="secondary"
          onClick={[Function]}
        >
          <ForwardRef(IonIcon)
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Add</title><path stroke-linecap='round' stroke-linejoin='round' d='M256 112v288M400 256H112' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
        </ForwardRef(IonFabButton)>
      </ForwardRef(IonFab)>
      <ForwardRef
        duration={5000}
        isOpen={true}
        message="Loading..."
        onDidDismiss={[Function]}
        spinner="crescent"
      />
      <ForwardRef
        backdropDismiss={false}
        buttons={
          Array [
            Object {
              "handler": [Function],
              "text": "OK",
            },
          ]
        }
        header=""
        isOpen={false}
        message=""
      />
    </ForwardRef(IonContent)>
  </ForwardRef(IonPage)>,
]
`;
